src/query/ts/m3db/encoded_unconsolidated_step_iterator.go

```diff
@@ -21,99 +21,95 @@
 	"github.com/m3db/m3/src/query/storage"
 	xts "github.com/m3db/m3/src/query/ts"
 )
 
 type encodedStepIterUnconsolidated struct {
-	mu          sync.RWMutex
-	lastBlock   bool
-	exhausted   bool
-	err         error
-	meta        block.Metadata
-	validIters  []bool
-	seriesMeta  []block.SeriesMeta
-	seriesIters []encoding.SeriesIterator
+	lastBlock      bool
+	idx            int
+	err            error
+	meta           block.Metadata
+	seriesMeta     []block.SeriesMeta
+	seriesIters    []encoding.SeriesIterator
+	expandedSeries [][]xts.Datapoints
 }
 
-func (it *encodedStepIterUnconsolidated) Current() (
-	block.UnconsolidatedStep,
+func (b *encodedBlockUnconsolidated) StepIter() (
+	block.UnconsolidatedStepIter,
 	error,
 ) {
-	it.mu.RLock()
-	if it.exhausted {
-		it.mu.RUnlock()
-		return nil, fmt.Errorf("out of bounds")
-	}
+	return &encodedStepIterUnconsolidated{
+		idx:            -1,
+		meta:           b.meta,
+		seriesMeta:     b.seriesMetas,
+		seriesIters:    b.seriesBlockIterators,
+		lastBlock:      b.lastBlock,
+		expandedSeries: make([][]xts.Datapoints, len(b.seriesBlockIterators)),
+	}, nil
+}
 
-	if it.err != nil {
-		it.mu.RUnlock()
-		return nil, it.err
+func (it *encodedStepIterUnconsolidated) Current() block.UnconsolidatedStep {
+	stepTime, _ := it.meta.Bounds.TimeForIndex(it.idx)
+	stepValues := make([]xts.Datapoints, len(it.expandedSeries))
+	for j, series := range it.expandedSeries {
+		stepValues[j] = series[it.idx]
 	}
 
-	var t time.Time
-	values := make([]xts.Datapoints, len(it.seriesIters))
+	step := storage.NewUnconsolidatedStep(stepTime, stepValues)
+	return step
+}
+
+func (it *encodedStepIterUnconsolidated) decodeSeries() bool {
+	values := make(xts.Datapoints, 0, initBlockReplicaLength)
 	for i, iter := range it.seriesIters {
-		// Skip this iterator if it's not valid
-		if !it.validIters[i] {
-			continue
+		values = values[:0]
+		for iter.Next() {
+			dp, _, _ := iter.Current()
+			values = append(values,
+				xts.Datapoint{
+					Timestamp: dp.Timestamp,
+					Value:     dp.Value,
+				})
 		}
 
-		dp, _, _ := iter.Current()
-		if i == 0 {
-			t = dp.Timestamp
+		if it.err = iter.Err(); it.err != nil {
+			return false
 		}
 
-		values[i] = []xts.Datapoint{{
-			Timestamp: dp.Timestamp,
-			Value:     dp.Value,
-		}}
+		it.expandedSeries[i] = values.AlignToBounds(it.meta.Bounds)
 	}
 
-	step := storage.NewUnconsolidatedStep(t, values)
-	it.mu.RUnlock()
-	return step, nil
+	return true
 }
 
 func (it *encodedStepIterUnconsolidated) Next() bool {
-	it.mu.Lock()
-	defer it.mu.Unlock()
-	if it.exhausted {
+	if it.err != nil {
 		return false
 	}
 
-	var anyNext bool
-	for i, iter := range it.seriesIters {
-		if iter.Next() {
-			anyNext = true
-			it.validIters[i] = true
-		} else {
-			it.validIters[i] = false
-		}
-
-		if err := iter.Err(); err != nil {
-			it.err = err
+	it.idx++
+	if it.idx == 0 {
+		// decode the series on initial Next() call
+		if success := it.decodeSeries(); !success {
```

Comment:
Am I reading correctly that this change switches us back to eager decompression? What's the reasoning behind the switch back?

Reply:
Yeah, so the original implementation was incorrect, so had to fix it. I had a few questions with unconsolidated step iteration as a concept, since it's a bit odd, and was thinking of maybe just getting rid of it in general.

Talked offline and will fix this up to do it lazily the right way instead.